zkevm-circuits/src/evm_circuit/op_execution/arithmetic.rs

```diff
@@ -0,0 +1,564 @@
+use super::super::{
+    BusMappingLookup, Case, Cell, Constraint, CoreStateInstance, ExecutionStep,
+    Lookup, Word,
+};
+use super::{CaseAllocation, CaseConfig, OpExecutionState, OpGadget};
+use halo2::{
+    arithmetic::FieldExt,
+    circuit::Region,
+    plonk::{Error, Expression},
+};
+use std::{array, convert::TryInto};
+
+#[derive(Clone, Debug)]
+struct AddSuccessAllocation<F> {
+    selector: Cell<F>,
+    swap: Cell<F>,
+    a: Word<F>,
+    b: Word<F>,
+    c: Word<F>,
+    carry: [Cell<F>; 32],
+}
+
+#[derive(Clone, Debug)]
+pub struct AddGadget<F> {
+    success: AddSuccessAllocation<F>,
```

Comment:
To me this should also contain the data of one or multiple `ExecutionStep`s which contain in this case an `ADD` instruction.

Reply:
Or should have implemented a way on which: `Provided a slice of `ExecutionStep`s related to the addition, witness all of the data and build all the lookups/constraints.

Reply:
But the lookups and constraints should not depend on proving time witness, it's fixed when keygen, right?

Reply:
The values you place on the lookups should no??

I think that you got a point in the sense iof that this just states the logic on which the allocation is done, not the actual witness-related actions.

Reply:
Yes, but the witness will be fed in fn `assign_*`, the gadget only needs to memorize which cell to assign it, right?

Reply:
You're absolutely right. Sorry.